### ISSUE TITLE
Cloudflare Unban Rule Malfunction

### ISSUE THREAD

**Issue by RaidForums** [fail2ban/fail2ban#2128]
**I am running the latest cloudflare action file updated 9 months ago but I am having a problem which is the unban rule no longer works. The get CURL request within this unban rule that is supposed to return 1 result for the script to be able to unban the IP instead returns all of page 1 so if you have multiple blocked IPs fail2ban is unable to unban the IP.**

I have about 1300 IPs on my Cloudflare Access Rules right now blocked because the script is no longer working as intended this should be a pretty easy fix.

**Comment by RaidForums** [comment]
Sorry, I originally misclicked without leaving an issue comment.

**Comment by RaidForums** [comment]
After looking into this I realise that no matter what I set the per_page value to on the API It is being reset to 25 when the request is fulfilled by the API. I have made a support ticket with Cloudflare for them to fix this.

**Comment by RaidForums** [comment]
The Cloudflare API has been updated and is working fine but the current unban action is still not working.

**Comment by RaidForums** [comment]
I am currently just making fail2ban run a php script I created to unban the IPs because I didn't see any simple and basic way of fixing the current problem. I guess i'll be waiting until one of the fail2ban developers fix this problem.

**Comment by wpkc** [comment]
My update to the cloudflare action works fine. <https://github.com/wpkc/fail2ban-action-cloudflare-restv4>

I think the use of `cut -d '"' -f6` would be problematic because there is no guarantee that the format of the JSON data returned will always have the IP address in the 6th delimited field.  

My method may be a bit more resource intensive, but I use a single line python script to extract the IP address from the first result in the JSON response data. Then I pipe it to xargs to build and execute the curl statement. I also use the `notes` parameter to make sure Fail2Ban is only deleting firewall rules that it created.  Otherwise, an admin or other program might have set it's own firewall rule only to have Fail2Ban delete it.

```
actionban = curl -s -X POST https://api.cloudflare.com/client/v4/user/firewall/access_rules/rules \
	-H "X-Auth-Email: <cfuser>" -H "X-Auth-Key: <cfkey>" -H "Content-Type: application/json" \
--data '{"mode":"block","configuration":{"target":"ip","value":"<ip>"},"notes":"Banned by Fail2Ban"}'

actionunban = curl -s -X GET -H "X-Auth-Email: <cfuser>" -H "X-Auth-Key: <cfkey>" -H "Content-Type: application/json" \
	"https://api.cloudflare.com/client/v4/user/firewall/access_rules/rules?page=1&per_page=5&mode=block&configuration.target=ip&configuration.value=<ip>&notes=Banned by Fail2Ban&match=all&order=configuration.value&direction=desc" | \
	python -c "import sys, json; print json.load(sys.stdin)['result'][0]['id'];" | \
	xargs -I@@ curl -s -X DELETE -H "X-Auth-Email: <cfuser>" -H "X-Auth-Key: <cfkey>" -H "Content-Type: application/json" https://api.cloudflare.com/client/v4/user/firewall/access_rules/rules/@@
```
This has been working flawlessly for me for the past year.

**Comment by hyperexpert** [comment]
Is there a way to add date and time to the ban notes? I tried declaring date in the init section and calling it in the notes but that didnt work!

**Comment by wpkc** [comment]
Did you try passing it as a parameter in the jail.local actionban, then using that parameter in action.d/cloudflare-restv4.conf?

**Comment by hyperexpert** [comment]
I tried adding this to cloudflare.conf file
`cftimestamp = $(date +%Y%m%d%H%M%S)`

then changing my curl command to:
`curl -s -o /dev/null -X POST -H 'X-Auth-Email: <cfuser>' -H 'X-Auth-Key: <cftoken>' \
            -H 'Content-Type: application/json' -d '{ "mode": "challenge", "configuration": { "target": "ip", "value": "<ip>" },"notes":"<cftimestamp>"}' \
            https://api.cloudflare.com/client/v4/user/firewall/access_rules/rules`

But that doesn't work

**Comment by hyperexpert** [comment]
Also tried the answer here but that also doesnt work :(

https://www.spinics.net/lists/fail2ban/msg00798.html

**Comment by sebres** [comment]
Well, fail2ban does not execute each parameter that you specified in shell, so this will substitute into:
```
curl ... -d '{ ... ,"notes":"$(date +%Y%m%d%H%M%S)"}' ...
```
Thus it would not work, because of the single quotation (no interpolation in shell).
Just compare in the shell:
```diff
-echo '{"test": "$(date +%Y%m%d%H%M%S)"}'
+echo "{\"test\": \"$(date +%Y%m%d%H%M%S)\"}"
```
You have to use double quotation mark (and to escape all others chars), or try to use something like this:
```diff
-curl ... -d '{ "mode": "challenge", "configuration": { "target": "ip", "value": "<ip>" },"notes":"<cftimestamp>"}' ...
+curl ... -d `printf '{ "mode": "challenge", "configuration": { "target": "ip", "value": "<ip>" },"notes":"%s"}' $(date +%Y%m%d%H%M%S)` ...
```

**Comment by hyperexpert** [comment]
@sebres thank you so much for replying back to me but unfortunately both of your examples did not work

I tried both example and both cause fail2ban to not start

```bash
curl -s -o /dev/null -X POST -H 'X-Auth-Email: <cfuser>' -H 'X-Auth-Key: <cftoken>' \
            -H 'Content-Type: application/json' -d "{ \"mode\": \"challenge\", \"configuration\": { \"target\": \"ip\", \"value\": \"<ip>\" },\"notes\": \"$(date +%Y%m%d%H%M%S)\" }" \
            https://api.cloudflare.com/client/v4/user/firewall/access_rules/rules
```

```bash
curl -s -o /dev/null -X POST -H 'X-Auth-Email: <cfuser>' -H 'X-Auth-Key: <cftoken>' \
            -H 'Content-Type: application/json' -d `printf '{ "mode": "challenge", "configuration": { "target": "ip", "value": "<ip>" },"notes":"%s"}' $(date +%Y%m%d%H%M%S)` \
            https://api.cloudflare.com/client/v4/user/firewall/access_rules/rules
```

**Comment by sebres** [comment]
> both cause fail2ban to not start

Well, although "does not start" is not good description of issue, but I think it is...

Because you should always escape `%` char with `%%` in fail2ban (python) config-files:
```diff
-%Y%m%d%H%M%S
+%%Y%%m%%d%%H%%M%%S
```
Anyway, next time please provide an error message (or log-excerpt) instead of "doesn't work".

**Comment by hyperexpert** [comment]
This worked great! Thank you so much for all the help.

**Comment by antfarmer** [comment]
I think I found the problem with the original issue of unbanning, at least for me. There was whitespace before the id in the result, so the url was really malformed, hence the 400 Bad request. All I did was add a little sed action to trim the id (no need for pulling in python): ` | sed -e 's/^[[:space:]]*//' -e 's/[[:space:]]*$//'`

Altogether: 

````
actionunban = curl -s -X DELETE "https://api.cloudflare.com/client/v4/user/firewall/access_rules/rules/$( \
              curl -s -X GET "https://api.cloudflare.com/client/v4/user/firewall/access_rules/rules?mode=block&configuration_target=ip&configuration_value=<ip>&page=1&per_page=1&match=all" \
              -H "X-Auth-Email: <cfuser>" \
              -H "X-Auth-Key: <cftoken>" \
              -H "Content-Type: application/json" | awk -F"[,:}]" '{for(i=1;i<=NF;i++){if($i~/'id'\042/){print $(i+1)}}}' | tr -d '"' | head -n 1 | sed -e 's/^[[:space:]]*//' -e 's/[[:space:]]*$//')" \
              -H "X-Auth-Email: <cfuser>" \
              -H "X-Auth-Key: <cftoken>" \
              -H "Content-Type: application/json"
````

Works for me now. I did have to write a script to delete all the old orphan rules though which took forever, as this must have been going on for a long time...

**Comment by nextend** [comment]
I had the same issue.

`curl -s -X GET -H 'X-Auth-Email: ****' -H 'X-Auth-Key: ***' 'https://api.cloudflare.com/client/v4/user/firewall/access_rules/rules?mode=block&configuration_target=ip&configuration_value=50.50.50.50&page=1&per_page=1'
`
Response:
```
{
  "result": [
    {
      "id": "88aa292f*****c970c7b817cb875",
      "paused": false,
      "modified_on": "2019-01-23T09:24:49.982274Z",
      "allowed_modes": [
        "whitelist",
        "block",
        "challenge",
        "js_challenge"
      ],
      "mode": "block",
      "notes": "",
      "configuration": {
        "target": "ip",
        "value": "50.50.50.50"
      },
      "scope": {
        "id": "2bb43143c3*****bfd1c8f6add5da",
        "email": "*****",
        "type": "user"
      },
      "created_on": "2019-01-23T09:24:49.984163142Z"
    }
  ],
  "success": true,
  "errors": [],
  "messages": [],
  "result_info": {
    "page": 1,
    "per_page": 1,
    "count": 1,
    "total_count": 1,
    "total_pages": 1
  }
}
```

Probably the `| cut -d'"' -f6` as it should on the response. So I installed jq and I find the id with the following unban action:

```
actionunban = curl -s -o /dev/null -X DELETE -H 'X-Auth-Email: <cfuser>' -H 'X-Auth-Key: <cftoken>' \
               https://api.cloudflare.com/client/v4/user/firewall/access_rules/rules/$(curl -s -X GET -H 'X-Auth-Email: <cfuser>' -H 'X-Auth-Key: <cftoken>' \
               'https://api.cloudflare.com/client/v4/user/firewall/access_rules/rules?mode=block&configuration_target=ip&configuration_value=<ip>&page=1&per_page=1' | jq -r '.r    esult[0].id')
```

**Comment by acs-ferreira** [comment]
Also have a look at https://www.vcloudnine.de/cloudflare-api-v4-and-fail2ban-fixing-the-unban-action/

**Comment by momala454** [comment]
do you have any update ? it's still broken  as of now

**Comment by sebres** [comment]
Merged in #2140 now (as working with and without `jq`).
